src/stats.ts

```diff
@@ -0,0 +1,121 @@
+import { Client } from './client';
+import { GenericObject, StatsEvent } from './types';
+import { isDfid } from './util/is';
+
+export class StatsClient extends Client {
+  /**
+   * Registers a new session with Doofinder, and returns
+   * the session ID generated for this session
+   *
+   * @param   {String}     sessionId   The sessionId
+   *
+   * @param   {Function}   callback    Optional, callback after everything works
+   *
+   * @return  {String}     The newly generated Session ID
+   */
+  public async requestSession(sessionId: string, hashid?: string): Promise<GenericObject> {
+    if (hashid) {
+      this.hashid = hashid;
+    }
+    return await this.stats(StatsEvent.Init, { session_id: sessionId, hashid: this.hashid });
+  }
+
+  /**
+   * Registers a click within this session, using the dfid
+   * or the database id and datatype.
+   *
+   * @param  {String}   sessionId           The sessionId
+   *
+   * @param  {String}   id                  Id of the result or Doofinder's internal ID
+   *                                        for the result.
+   *
+   * @param  {String}   datatype            Optional. If the id is not a Doofinder id
+   *                                        this argument is required.
+   *
+   * @param  {String}   query               Optional. Search terms.
+   *
+   * @param  {String}   custom_results_id   Optional. Id of the custom results
+   *                                        that produced the current set of
+   *                                        results, including the current one
+   *                                        being clicked.
+   *
+   */
+  public async registerClick(
+    sessionId: string,
+    id: string,
+    datatype?: string,
+    query?: string,
+    customResultsId?: string | number
+  ): Promise<GenericObject> {
```

Comment:
I would use object destructuring here too… missing values will be undefined. Again, hashid will be added by the client if not present. You can make the params explicit, I mean:

```typescript
export interface ClickParams extends BaseParams { /* … */}
public async registerClick({
  sessionId, 
  hashid, 
  dfid,
  id,
  datatype,
  query,
  customResultsId
}: ClickParams): Promise<GenericObject> {
  /* … */
}
```

and check:

```js
if (datatype && id) {
  // use them
} else if (dfid && isDfid(dfid)) {
  // use it
} else {
  // throw saying that datatype + id, or a valid dfid are required
}
```

As mentioned in other PR, undefined values should not be added to the querystring (null values should) so you reduce checks to the minimum, just forward parameters.

You can also infer hashid from dfid.

Reply:
If we leave the object-as-parameter discussion aside, the TypeScript parameter definition system works quite alright, having the optional parameters as `undefined` too just like the object destructuring (it wouldn't surprise me to find it does something alike in the generated code). So... I don't see the improvement here.

Reply:
Explaining the improvement is quite long to explain and would take an entire article, so… [here you are one](https://levelup.gitconnected.com/always-pass-one-argument-to-your-javascript-function-4140d909937e).